File: clients/swift/Tests/QuidnugTests/QuidTests.swift
import XCTest
@testable import Quidnug

final class QuidTests: XCTestCase {
    func testGenerateHasExpectedIdFormat() throws {
        let q = Quid.generate()
        XCTAssertEqual(q.id.count, 16)
        XCTAssertTrue(q.hasPrivateKey)
        XCTAssertNotNil(q.publicKeyHex)
        XCTAssertNotNil(q.privateKeyHex)
    }

    func testSignVerifyRoundtrip() throws {
        let q = Quid.generate()
        let sig = try q.sign(Data("hello".utf8))
        XCTAssertTrue(q.verify(Data("hello".utf8), sigHex: sig))
        XCTAssertFalse(q.verify(Data("tampered".utf8), sigHex: sig))
    }

    func testPrivateHexRoundtrip() throws {
        let q = Quid.generate()
        let r = try Quid.fromPrivateHex(q.privateKeyHex!)
        XCTAssertEqual(q.id, r.id)
        XCTAssertEqual(q.publicKeyHex, r.publicKeyHex)
        XCTAssertTrue(r.hasPrivateKey)
    }

    func testReadOnlyCannotSign() throws {
        let q = Quid.generate()
        let ro = try Quid.fromPublicHex(q.publicKeyHex)
        XCTAssertFalse(ro.hasPrivateKey)
        XCTAssertEqual(q.id, ro.id)
        XCTAssertThrowsError(try ro.sign(Data("x".utf8)))
    }

    func testCrossVerifyFails() throws {
        let a = Quid.generate()
        let b = Quid.generate()
        let sig = try a.sign(Data("shared".utf8))
        XCTAssertFalse(b.verify(Data("shared".utf8), sigHex: sig))
    }
}

File: clients/dotnet/tests/QuidTests.cs
using System.Text;
using Xunit;

namespace Quidnug.Client.Tests;

public class QuidTests
{
    [Fact]
    public void GenerateHasExpectedIdFormat()
    {
        using var q = Quid.Generate();
        Assert.Equal(16, q.Id.Length);
        Assert.True(q.HasPrivateKey);
        Assert.NotNull(q.PublicKeyHex);
        Assert.NotNull(q.PrivateKeyHex);
    }

    [Fact]
    public void SignVerifyRoundtrip()
    {
        using var q = Quid.Generate();
        byte[] data = Encoding.UTF8.GetBytes("hello");
        string sig = q.Sign(data);
        Assert.True(q.Verify(data, sig));
        Assert.False(q.Verify(Encoding.UTF8.GetBytes("tampered"), sig));
    }

    [Fact]
    public void PrivateHexRoundtrip()
    {
        using var q = Quid.Generate();
        Assert.NotNull(q.PrivateKeyHex);
        using var r = Quid.FromPrivateHex(q.PrivateKeyHex!);
        Assert.Equal(q.Id, r.Id);
        Assert.Equal(q.PublicKeyHex, r.PublicKeyHex);
    }

    [Fact]
    public void ReadOnlyQuidCannotSign()
    {
        using var q = Quid.Generate();
        using var ro = Quid.FromPublicHex(q.PublicKeyHex);
        Assert.False(ro.HasPrivateKey);
        Assert.Equal(q.Id, ro.Id);
        Assert.Throws<InvalidOperationException>(
            () => ro.Sign(Encoding.UTF8.GetBytes("x")));
    }

    [Fact]
    public void CrossVerifyBetweenQuidsFails()
    {
        using var a = Quid.Generate();
        using var b = Quid.Generate();
        string sig = a.Sign(Encoding.UTF8.GetBytes("shared"));
        Assert.False(b.Verify(Encoding.UTF8.GetBytes("shared"), sig));
    }
}
